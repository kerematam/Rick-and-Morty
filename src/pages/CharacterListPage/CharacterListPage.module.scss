.root {
  flex-grow: 1;
}

.progress_bar {
  position: fixed;
  z-index: 100;
  top: 0;
  right: 0;
  width: 100%;
}

.reload_button {
  padding: 10px;
  margin: 20px auto 20px auto;
  width: 100%;
}
