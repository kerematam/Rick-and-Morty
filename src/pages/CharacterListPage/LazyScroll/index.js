import React from 'react'
import Grid from '@material-ui/core/Grid'
import InfiniteScroll from 'react-infinite-scroller'
import Card from './CharacterCard'

const LazyScroll = ({ characters = [], loadMore, loading, hasMore }) => {
  return (
    <InfiniteScroll pageStart={0} loadMore={loadMore} hasMore={true}>
      <Grid container spacing={3}>
        {characters.map(character => (
          <Grid key={character.id} item xs={12} sm={6} md={4} lg={3}>
            <Card character={character} />
          </Grid>
        ))}
      </Grid>
    </InfiniteScroll>
  )
}

export default React.memo(LazyScroll)
