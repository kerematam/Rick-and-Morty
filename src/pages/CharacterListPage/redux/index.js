import { useInjectReducer, useInjectSaga } from 'redux-injectors'
import { createSlice } from '@reduxjs/toolkit'
import saga from './sagas'

export const NAMESPACE = 'CharacterListPage'

export const initialState = {
  loading: false,
  page: 1,
  hasMore: true,
  characters: [],
}

const slice = createSlice({
  name: NAMESPACE,
  initialState,
  reducers: {
    loadCharacters: state => {
      state.loading = true
    },
    loadCharactersSucceed: (state, action) => {
      state.characters = action.payload.characters
      state.page = action.payload.page
      state.hasMore = action.payload.hasMore
      state.loading = false
      state.error = false
    },
    loadCharactersFailed: (state, action) => {
      state.loading = false
      state.error = true
    },
  },
})

export const { reducer, actions } = slice

//  Injectors
export const useCharacterReducerInjector = () =>
  useInjectReducer({ key: NAMESPACE, reducer })
export const useCharacterSagaInjector = () =>
  useInjectSaga({ key: NAMESPACE, saga })
export const useCharacterReduxInjector = () => {
  useCharacterReducerInjector()
  useCharacterSagaInjector()
}

// Selectors
export const characterSelector = state => state[NAMESPACE]
