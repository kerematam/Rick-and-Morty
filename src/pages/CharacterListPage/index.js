import React, { useCallback } from 'react'
import Button from '@material-ui/core/Button'
import { Helmet } from 'react-helmet-async'
import { useSelector, useDispatch } from 'react-redux'
import { actions, useCharacterReduxInjector, characterSelector } from './redux'
import LazyScroll from './LazyScroll'
import styles from './CharacterListPage.module.scss'

export default React.memo(function CharacterListPage() {
  useCharacterReduxInjector()
  const dispatch = useDispatch()
  const { loading, error, characters, hasMore } = useSelector(characterSelector)
  const loadMore = useCallback(() => {
    if (!loading && !error && hasMore) {
      dispatch(actions.loadCharacters())
    }
  }, [loading, error, hasMore, dispatch])

  return (
    <div className={styles.root}>
      <Helmet>
        <title>Rick and Morty</title>
      </Helmet>
      {error && (
        <Button onClick={loadMore}>
          Could not fetch data; Try to reload again.
        </Button>
      )}
      <LazyScroll hasMore loading characters={characters} loadMore={loadMore} />
    </div>
  )
})
