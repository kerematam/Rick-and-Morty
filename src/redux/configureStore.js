import { createStore, applyMiddleware, compose } from 'redux'
import createSagaMiddleware from 'redux-saga'
import createReducer from './rootReducers'
import { routerMiddleware } from 'connected-react-router'
import { createInjectorsEnhancer, forceReducerReload } from 'redux-injectors'

function configureStore(initialState = {}, history) {
  let composeEnhancers = compose
  const reduxSagaMonitorOptions = {}

  if (process.env.NODE_ENV !== 'production' && typeof window === 'object') {
    if (window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__)
      composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__({})
  }

  const sagaMiddleware = createSagaMiddleware(reduxSagaMonitorOptions)
  const { run: runSaga } = sagaMiddleware
  const middlewares = [sagaMiddleware, routerMiddleware(history)]

  const enhancers = [
    applyMiddleware(...middlewares),
    createInjectorsEnhancer({
      createReducer,
      runSaga,
    }),
  ]

  const store = createStore(
    createReducer(),
    initialState,
    composeEnhancers(...enhancers)
  )

  if (module.hot) {
    module.hot.accept('./rootReducers', () => {
      forceReducerReload(store)
    })
  }

  return store
}

export default configureStore
