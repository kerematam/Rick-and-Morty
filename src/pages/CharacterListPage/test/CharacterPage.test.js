import React, { useCallback } from 'react'
import renderer from 'react-test-renderer'
import CharacterListPage from '../index'
import { Provider } from 'react-redux'
import configureStore from 'redux/configureStore'
import { HelmetProvider } from 'react-helmet-async'

const setup = () => {
  const store = configureStore()
  const wrapper = renderer
    .create(
      <Provider store={store}>
        <HelmetProvider>
          <CharacterListPage />
        </HelmetProvider>
      </Provider>
    )
    .toJSON()

  return {
    wrapper,
  }
}

describe('CharacterListPage', () => {
  test('should render correctly', () => {
    const { wrapper } = setup()
    expect(wrapper).toMatchSnapshot()
  })
})
