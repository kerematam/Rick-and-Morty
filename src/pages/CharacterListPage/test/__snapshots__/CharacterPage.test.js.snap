// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CharacterListPage should render correctly 1`] = `
<div
  className="root"
>
  <div>
    <div
      className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-3"
    />
  </div>
</div>
`;
